crates/oxc_linter/src/rules/eslint/no_control_regex.rs

```diff
@@ -0,0 +1,351 @@
+use lazy_static::lazy_static;
+use oxc_ast::{
+    ast::{Argument, Expression, RegExpFlags},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{Atom, GetSpan, Span};
+use regex::{Matches, Regex};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-control-regex): Unexpected control character(s)")]
+#[diagnostic(
+    severity(error),
+    help("Unexpected control character(s) in regular expression: \"{0}\"")
+)]
+struct NoControlRegexDiagnostic(Atom, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoControlRegex;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallows control characters and some escape sequences that match
+    /// control characters in regular expressions.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Control characters are special, invisible characters in the ASCII range
+    /// 0-31. These characters are rarely used in JavaScript strings so a
+    /// regular expression containing elements that explicitly match these
+    /// characters is most likely a mistake.
+    ///
+    /// ### Example
+    ///
+    /// Examples of **incorrect** code for this rule:
+    ///
+    /// ```javascript
+    /// var pattern1 = /\x00/;
+    /// var pattern2 = /\x0C/;
+    /// var pattern3 = /\x1F/;
+    /// var pattern4 = /\u000C/;
+    /// var pattern5 = /\u{C}/u;
+    /// var pattern6 = new RegExp("\x0C"); // raw U+000C character in the pattern
+    /// var pattern7 = new RegExp("\\x0C"); // \x0C pattern
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    ///
+    /// ```javascript
+    /// var pattern1 = /\x20/;
+    /// var pattern2 = /\u0020/;
+    /// var pattern3 = /\u{20}/u;
+    /// var pattern4 = /\t/;
+    /// var pattern5 = /\n/;
+    /// var pattern6 = new RegExp("\x20");
+    /// var pattern7 = new RegExp("\\t");
+    /// var pattern8 = new RegExp("\\n");
+    /// ```
+    NoControlRegex,
+    correctness
+);
+
+impl Rule for NoControlRegex {
+    fn run<'a>(&self, node: &AstNode<'a>, context: &LintContext<'a>) {
+        if let Some((pattern, flags, span)) = regex_pattern(node) {
+            let mut violations: Vec<&str> = Vec::with_capacity(usize::min(4, pattern.len()));
```

Comment:
What's your rationale for `with_capacity`? We won't be getting any errors for the common case so this is allocating more than necessary?

Reply:
I'm attempting to avoid unnecessary reallocs. Really I should try a lazily initialized static vec and just re-use that.